File: lib/service/money_service.rb
require_relative '../model/types/result'
require_relative '../model/money'

def print_money(money)
  puts "Money:"
  puts "\tsold: #{money.sold}"
end

def deposit(money)
  print "Money to deposit: "
  input_str = gets.chomp

  begin
    input_int = Integer(input_str)
    money.sold = input_int
    Result.new(true, "Deposit successfully", money)
  rescue ArgumentError, TypeError
    Result.new(false, "Invalid Input - retry", input_str)
  end
end

def withdrawal(money)
  print "Amount to withdraw: "
  input_str = gets.chomp

  begin
    input_int = Integer(input_str)
    if input_int > money.sold
      Result.new(false, "Insufficient balance", input_int)
    else
      money.sold -= input_int
      Result.new(true, "Withdrawal successfully", money)
    end
  rescue ArgumentError, TypeError
    Result.new(false, "Invalid Input - retry", input_str)
  end
end

File: src/service/money_service.py
from src.model.money import Money
from src.model.type.result import Result


def print_money(money: Money) -> None:
    print(f"money:\n  sold: {money.sold}")


def deposit(money: Money) -> Result:
    input_str: str = input("Money to deposit: ")
    try:
        input_int: int = int(input_str)
        money.sold = money.sold + input_int
        return Result(True, "Deposit successfully", money)
    except ValueError:
        return Result(False, "Invalid Input - retry", input_str)


def withdrawal(money: Money) -> Result:
    input_str: str = input("Amount to withdraw: ")
    try:
        input_int: int = int(input_str)
        if input_int > money.sold:
            return Result(False, "Insufficient balance", input_int)
        money.sold = money.sold - input_int
        return Result(True, "Withdrawal successfully", money)
    except ValueError:
        return Result(False, "Invalid Input - retry", input_str)
